compiler/modules/bir/bir.bal

```diff
@@ -584,30 +584,81 @@ public type EqualityInsn readonly & record {|
     Operand[2] operands;
 |};
 
-# Call a function.
-# This is a not a terminator.
-# This is a PPI. A panic in the called function
+public type FunctionOperand FunctionConstOperand|Register;
+
+# The common supertype of the two call instructions.
+# The following applies to both instructions.
+# It is not a terminator.
+# It will be a PPI. A panic in the called function
 # goes to the onPanic label in the basic block.
+# (This isn't implemented yet, since we haven't implemented `trap`.)
 # Regardless of where the function itself panics,
 # any function call could result in a stack overflow panic.
-# XXX This does not handle functions that don't return
-# (i.e. with return type of never)
-public type CallInsn readonly & record {|
+# It can also panic due to memory allocation for uniform function call
+# (Panics due to memory allocation are not handled gracefully)
+#
+# The first operand is the function value to be called;
+# this operand must be a subtype of the function basic type.
+# The remaining operands are the arguments.
+# A function value has an inherent type, which is determined
+# by the declared parameter types and return type; these types are
+# called atomic types. In an atomic function type, the type of the parameters
+# of the function is represented as a tuple type; when the function is declared
+# with a rest parameter, the last member descriptor in the tuple type
+# will be a rest descriptor.
+# In general, function types can be unions and intersections of atomic function types.
+# The semantics of function call in Ballerina are that the arguments
+# are a list. Static type checking of a call ensures that the list of arguments passed to
+# a function value belongs to the tuple type for the parameters of the inherent type
+# of the function value.
+# In all cases, it is the responsibility of the caller to deal with rest arguments
+# in the call expression, by splicing the rest argument into the list of all the arguments;
+# the call instructions assume this splicing has already taken place.
+# In the normal case, it is similarly the responsibility of the callee to handle
+# rest parameters in the function definition, by constructing a list from the corresponding
+# trailing part of the list of all the arguments, before binding this list to the rest parameter.
+# However, in some cases it is possible to shortcut the callee's handling of the rest
+# parameter, by having the caller construct the list to be bound to the rest parameter;
+# we call this shortcut `restParamIsList`.
+# The two instructions differ in how they distinguish the case `restParamIsList` is in effect.
```

Comment:
A `t:FunctionSignature` represents the parameter types after the list has been constructed for the rest parameter.